### PR TITLE
Reject orders with receivers as SC wallets

### PR DESCRIPTION
**Pull request by nlordell**
This PR rejects orders with a receiver address as an SC wallet when buying ETH.

This should be removed once we implement access list support and SC wallets can be supported.

Created #685 to capture the work required to support this properly with the current contracts.

### Test Plan

<details><summary>Did some manual testing.</summary>

After spinning up a local orderbook, find a SC that can receive ETH (I just used the GPv2 Settlement contract as if it were  an SC wallet :stuck_out_tongue:). Verify that the address has code:
```
$ curl -s 'http://localhost:8545' \
  -H 'Content-Type: application/json' \
  --data-raw '{"jsonrpc":"2.0","id":1,"method":"eth_getCode","params":["0x9561c133dd8580860b6b7e504bc5aa500f0f06a7","latest"]}' \
  | jq '.result | length'
22352
```

Now author an order that transfers to the SC and see that it succeeds.
```
$ curl -s 'http://localhost:8080/api/v1/orders' \
  -H 'Content-Type: application/json' \
  --data-raw "$(cat << EOF
    {
      "sellToken": "0xe78a0f7e598cc8b0bb87894b0f60dd2a88d6a8ab",
      "buyToken": "0xe78a0f7e598cc8b0bb87894b0f60dd2a88d6a8ab",
      "receiver": "0x9561c133dd8580860b6b7e504bc5aa500f0f06a7",
      "sellAmount": "1234567890",
      "buyAmount": "1234567890",
      "validTo": 1622725574,
      "appData": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "feeAmount": "1234568890",
      "kind": "buy",
      "partiallyFillable": true,
      "signature": "0xdaec39dbdce62a1f4ed138cab5333a358277373debd71e4b52aeae9e807bac58241ceca6730bf9ebd13314424ea3e787d841c101c56e3f090d6474008b9f5efa1c",
      "signingScheme": "eip712",
      "from": "0x90F8bf6A479f320ead074411a4B0e7944Ea8c9C1"
    }
EOF
)"
"0x3cc0d4ec51ab940dcdec4c9b9a66c2fe28f7334195ca98227427cb2ced3332ce90f8bf6a479f320ead074411a4b0e7944ea8c9c160b8d3c6"
```

Author the same order, but instead buy native token and see that it fails with the expected message:
```
$ curl -s 'http://localhost:8080/api/v1/orders' \
  -H 'Content-Type: application/json' \
  --data-raw "$(cat << EOF
    {
      "sellToken": "0xe78a0f7e598cc8b0bb87894b0f60dd2a88d6a8ab",
      "buyToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
      "receiver": "0x9561c133dd8580860b6b7e504bc5aa500f0f06a7",
      "sellAmount": "1234567890",
      "buyAmount": "1234567890",
      "validTo": 1622726574,
      "appData": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "feeAmount": "1234568890",
      "kind": "buy",
      "partiallyFillable": true,
      "signature": "0x84d7cb9e0ccf2e2af129b23b03591f62ff5af0f5367c7f54e6e14946d4bae0197aa22192c5709581168bd80dcc5c3bda626ae37546cd0dfdb0108e3c025d3d201c",
      "signingScheme": "eip712",
      "from": "0x90F8bf6A479f320ead074411a4B0e7944Ea8c9C1"
    }
EOF
)" | jq
{
  "errorType": "TransferEthToContract",
  "description": "Setting receiver to a smart contract wallet when buying Ether is currently not supported"
}
```

</details>

### CODE REVIEW

**Review by bh2smith — APPROVED** [review]
Nice! I guess my test case came in handy!

File: orderbook/src/orderbook.rs [hunk 1]
```diff
@@ -118,6 +125,17 @@ impl Orderbook {
             return Ok(AddOrderResult::InsufficientFunds);
         }
 
+        if order.order_creation.buy_token == BUY_ETH_ADDRESS {
+            let code_size = self
+                .code_fetcher
+                .code_size(order.actual_receiver())
+                .await
+                .unwrap_or(1);
```
Comment (vkgnosis) [inline orderbook/src/orderbook.rs:133]: Why unwrap_or here? If fetching the code size fails shouldn't we bubble this up so it results in an internal server error?

**Review by vkgnosis — APPROVED** [review]

File: orderbook/src/orderbook.rs (hunk 1 above)
Comment (nlordell) [reply orderbook/src/orderbook.rs]: :see_no_evil: fixed!

File: orderbook/src/orderbook.rs [hunk 2]
```diff
@@ -32,6 +35,7 @@ pub enum AddOrderResult {
     InsufficientFunds,
     InsufficientFee,
     UnsupportedToken(H160),
+    TransferEthToContract,
```
Comment (fleupold) [inline orderbook/src/orderbook.rs:38]: Can you add documentation for this to open api please?

File: shared/src/code.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+use crate::Web3;
+use anyhow::Result;
+use ethcontract::H160;
+
+#[mockall::automock]
+#[async_trait::async_trait]
+pub trait CodeFetching: Send + Sync {
```
Comment (fleupold) [inline shared/src/code.rs:7]: nanonit: Maybe all these helper traits could be combined in a web3 module or the like?

**Review by fleupold — APPROVED** [review]
Good fix

File: shared/src/code.rs (hunk 3 above)
Comment (nlordell) [reply shared/src/code.rs:7]: As far as I can tell, this is the only one for now:
```
$ rg 'for Web3'
orderbook/src/account_balances.rs
140:impl BalanceFetching for Web3BalanceFetcher {

shared/src/code.rs
13:impl CodeFetching for Web3 {
```

I moved it to a `web3_traits` module that we can add to over time.

File: orderbook/openapi.yml [hunk 4]
```diff
@@ -561,14 +561,15 @@ components:
           type: string
           enum:
             [
+              WrongOwner,
```
Comment (bh2smith) [inline orderbook/openapi.yml:564]: What is the purpose of moving this to the top? I kind liked that it was in alphabetical order before.

Comment (nlordell) [reply orderbook/openapi.yml:564]: Sorry! I had copied it from the order in the `enum`. I'll change it back to alphabetical order.

**Pull request merged**